src/app/clusters/ecosystem-information-server/ecosystem-information-server.cpp

```diff
@@ -0,0 +1,375 @@
+                                                                                     uint64_t aUniqueLocationIdsLastEditEpochUs)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mUniqueLocationIds.push_back(std::move(aUniqueLocationId));
+    mUniqueLocationIdsLastEditEpochUs = aUniqueLocationIdsLastEditEpochUs;
+    return *this;
+}
+
+std::unique_ptr<EcosystemDeviceStruct> EcosystemDeviceStruct::Builder::Build()
+{
+    VerifyOrReturnValue(!mIsAlreadyBuilt, nullptr, ChipLogError(Zcl, "Build() already called"));
+    VerifyOrReturnValue(mDeviceName.size() <= kDeviceNameMaxSize, nullptr, ChipLogError(Zcl, "Device name too large"));
+    VerifyOrReturnValue(mOriginalEndpoint != kInvalidEndpointId, nullptr, ChipLogError(Zcl, "Invalid original endpoint"));
+    VerifyOrReturnValue(!mDeviceTypes.empty(), nullptr, ChipLogError(Zcl, "No device types added"));
+    VerifyOrReturnValue(mUniqueLocationIds.size() <= kUniqueLocationIdsListMaxSize, nullptr,
+                        ChipLogError(Zcl, "Too many location ids"));
+
+    for (auto & locationId : mUniqueLocationIds)
+    {
+        VerifyOrReturnValue(locationId.size() <= kUniqueLocationIdMaxSize, nullptr, ChipLogError(Zcl, "Location id too long"));
+    }
+
+    // std::make_unique does not have access to private constructor we workaround with using new
+    std::unique_ptr<EcosystemDeviceStruct> ret{ new EcosystemDeviceStruct(
+        std::move(mDeviceName), mDeviceNameLastEditEpochUs, mBridgedEndpoint, mOriginalEndpoint, std::move(mDeviceTypes),
+        std::move(mUniqueLocationIds), mUniqueLocationIdsLastEditEpochUs) };
+    mIsAlreadyBuilt = true;
+    return ret;
+}
+
+CHIP_ERROR EcosystemDeviceStruct::Encode(const AttributeValueEncoder::ListEncodeHelper & aEncoder, const FabricIndex & aFabricIndex)
+{
+    Structs::EcosystemDeviceStruct::Type deviceStruct;
+    if (!mDeviceName.empty())
+    {
+        deviceStruct.deviceName.SetValue(CharSpan(mDeviceName.c_str(), mDeviceName.size()));
+        // When there is a device name we also include mDeviceNameLastEditEpochUs
+        deviceStruct.deviceNameLastEdit.SetValue(mDeviceNameLastEditEpochUs);
+    }
+    deviceStruct.bridgedEndpoint  = mBridgedEndpoint;
+    deviceStruct.originalEndpoint = mOriginalEndpoint;
+    deviceStruct.deviceTypes = DataModel::List<const Structs::DeviceTypeStruct::Type>(mDeviceTypes.data(), mDeviceTypes.size());
+
+    std::vector<CharSpan> locationIds;
+    locationIds.reserve(mUniqueLocationIds.size());
+    for (auto & id : mUniqueLocationIds)
+    {
+        locationIds.push_back(CharSpan(id.c_str(), id.size()));
+    }
+    deviceStruct.uniqueLocationIDs = DataModel::List<CharSpan>(locationIds.data(), locationIds.size());
+
+    deviceStruct.uniqueLocationIDsLastEdit = mUniqueLocationIdsLastEditEpochUs;
+
+    // TODO(#33223) this is a hack, use mFabricIndex when it exists. Additionally checking fabric
+    // index matches should happen at the top of this method to prevent building the encodable
```

Comment:
Then you have to make assumptions here about which fields are sensitive and which are not, which can diverge from the actual spec/XML....

Reply:
I will update the comment.

Initially when I wrote that comment I assumed all read were fabric filtered read. I have since learned that we always have to encode